File: 0540-single-element-in-a-sorted-array/0540-single-element-in-a-sorted-array.cpp
class Solution {
public:
    int singleNonDuplicate(vector<int>& arr) {
            
        // Brut Solution 
        
//         int N = arr.size();
//         if(N == 1) 
//             return arr[0];
//         if(arr[0] != arr[1]) 
//             return arr[0];
//         if(arr[N-1] != arr[N-2]) 
//             return arr[N-1];
        
        
//         for(int i=1;i<N-1;i++) 
//         {
//             if(arr[i] != arr[i-1] && arr[i] != arr[i+1]) 
//                 return arr[i];
//         }
        
//         return -1;
        
        int N = arr.size();
        
        if(N == 1) 
            return arr[0];
        
        int low = 0, high = arr.size()-1;
        
        if(arr[low] != arr[low+1])
            return arr[low];
        if(arr[high] != arr[high-1]) 
            return arr[high];
        low++;
        high--;
        
        
        while(low <= high) 
        {
            int mid = (low + high)/2;
            
            if(arr[mid-1] != arr[mid] && arr[mid+1] != arr[mid]) 
                return arr[mid];
            
            // If mid is even && its pair exist on the left side means (odd,even) it means single element lies in right side part.
            if(mid%2 == 0 && arr[mid-1] == arr[mid])
            {
                // (odd, even) pair answer lies in left side.
                high = mid-1;
            }
            else if(mid%2 == 0 && arr[mid] == arr[mid+1])
            {   // (even,odd) pair so single element lies in right side.
                low = mid+1;
            }
            else if(mid%2 == 1 && arr[mid-1] == arr[mid])
            {
                // (even,odd) pair 
                low = mid+1;
            }
            else
                high = mid-1;
        }
        
        return -1;
    }
};

File: 0540-single-element-in-a-sorted-array/0540-single-element-in-a-sorted-array.java
class Solution {
    public int singleNonDuplicate(int[] nums) {
        int n = nums.length;

        if(n == 1) 
            return nums[0];
        if(nums[0] != nums[1]) 
            return nums[0];
        if(nums[n-1] != nums[n-2]) 
            return nums[n-1];

        int low = 1, high = n-2;
        // Approach is simple 

        //  left half --> even index (first appearance) odd index (second appearance).
        // (since single element break that chain) so -----> right half --> odd index (first appearance) even index (second appearance).
        while(low <= high) 
        {
            int mid = low + (high-low)/2;
            if(nums[mid] != nums[mid-1] && nums[mid] != nums[mid+1])
                return nums[mid];
            if(mid%2 == 0)
            {
                if(nums[mid] == nums[mid+1])
                    low = mid+1;
                else
                    high = mid-1;
            }
            else
            {
                if(nums[mid] == nums[mid-1])
                    low = mid+1;
                else
                    high = mid-1;
            }
        }
        return -1;
    }
}